File: app02_csharp_wiki_chat/aspnet-backend/SecureLearning.Backend/Translation/AiWorkerTranslationGateway.cs
namespace SecureLearning.Backend.Translation;

/// <summary>
/// Calls django-ai-worker's <c>POST /api/chat/translate</c>. Any failure (worker down,
/// network error, malformed response) is swallowed and logged — see <see cref="ITranslationGateway"/>.
/// </summary>
public sealed class AiWorkerTranslationGateway(HttpClient httpClient, ILogger<AiWorkerTranslationGateway> logger)
    : ITranslationGateway
{
    private sealed record TranslateRequest(string Text, string SourceLang, string TargetLang);
    private sealed record TranslateResponse(string? Translation);

    public async Task<string?> TranslateAsync(string text, string sourceLang, string targetLang,
        CancellationToken ct = default)
    {
        try
        {
            using var response = await httpClient.PostAsJsonAsync("/api/chat/translate",
                new TranslateRequest(text, sourceLang, targetLang), ct);
            response.EnsureSuccessStatusCode();
            var body = await response.Content.ReadFromJsonAsync<TranslateResponse>(cancellationToken: ct);
            return body?.Translation;
        }
        catch (Exception e) when (e is not OperationCanceledException)
        {
            logger.LogWarning(e, "AI worker translation of {Source} -> {Target} failed", sourceLang, targetLang);
            return null;
        }
    }
}

File: app_java_wiki_chat/spring-backend/src/main/java/com/platform/securelearning/translation/AiWorkerTranslationGateway.java
package com.platform.securelearning.translation;

import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.client.RestClient;

/**
 * Calls django-ai-worker's {@code POST /api/chat/translate}. Any failure (worker down,
 * network error, malformed response) is swallowed and logged — see {@link TranslationGateway}.
 */
public class AiWorkerTranslationGateway implements TranslationGateway {

    private static final Logger log = LoggerFactory.getLogger(AiWorkerTranslationGateway.class);

    private final RestClient restClient;

    public AiWorkerTranslationGateway(RestClient restClient) {
        this.restClient = restClient;
    }

    @Override
    public Optional<String> translate(String text, String sourceLang, String targetLang) {
        try {
            TranslateResponse response = restClient.post()
                    .uri("/api/chat/translate")
                    .body(new TranslateRequest(text, sourceLang, targetLang))
                    .retrieve()
                    .body(TranslateResponse.class);
            return response == null ? Optional.empty() : Optional.ofNullable(response.translation());
        } catch (Exception e) {
            log.warn("AI worker translation of {} -> {} failed", sourceLang, targetLang, e);
            return Optional.empty();
        }
    }
}
